### PR TITLE
Support __available__((swift_attr("@Sendable")))

### PR DESCRIPTION
**Pull request by beccadax**
This PR implements most of the sendability auditing support for ClangImporter. Specifically:

* `__attribute__((swift_attr("@Sendable")))` is now allowed on Clang decls, even when it would not be allowed on equivalent Swift decls.
* A Clang type declaration is imported with a synthesized `@unchecked Sendable` conformance if it:
    * Has `@Sendable`, but not plain `@_nonSendable`.
    * Has the `enum_extensibility`, `flag_enum`, `swift_newtype`, or `swift_wrapper` Clang attributes, but not plain `@_nonSendable`.
    * Has the `ns_error_domain` Clang attribute, **regardless of the presence** of plain `@_nonSendable`. (`Error` is always `Sendable`, so error codes must also be `Sendable`.)
* We now force implicit `Sendable` conformances before printing the contents of a module, and we print the extensions generated by `@_nonSendable` even if we would normally skip unavailable or implicit extensions.
* When ClangImporter is using a clang that supports https://reviews.llvm.org/D112773 (see apple/llvm-project#3547), it will define `__SWIFT_ATTR_SUPPORTS_SENDABLE_DECLS` to be `1`.

It includes a number of related refactorings:

* `SynthesizedProtocolAttr`s can now create `@unchecked` conformances.
* `SynthesizedFileUnit`s can now be attached to any `FileUnit`, not just `SourceFile`.
* `ClangImporter::Implementation::diagnose()` has been overloaded so that, if a `HeaderLoc` (wrapping a `clang::SourceLocation` and related info) is passed in place of a `SourceLoc`, it will emit a diagnostic in the header file. This replaces several lines of boilerplate per use site.

Fixes rdar://84431708. It does not implement two things that I will defer to 85569247:

* `@_nonSendable` in parameter position.
* Implicit `@Sendable` (probably `@_unsafeSendable` in practice) on completion handler parameters.

Some tests will not pass without https://reviews.llvm.org/D112773, which I am in the process of cherry-picking to our clang branches.

This should only cause language behavior changes if you consume headers that include the currently-invalid `__attribute__((swift_attr("@Sendable")))`, `__attribute__((swift_attr("@_nonSendable")))`, or  `__attribute__((swift_attr("@_nonSendable(_assumed)")))` attributes.

### CODE REVIEW

**Comment by beccadax** [comment]
With apple/llvm-project#3547

@swift-ci please smoke test

**Comment by beccadax** [comment]
With apple/llvm-project#3547

@swift-ci please smoke test

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -1769,6 +1769,24 @@ bool ShouldPrintChecker::shouldPrint(const Pattern *P,
...
+
+  auto nonSendable =
+      ED->getExtendedNominal()->getAttrs().getEffectiveSendableAttr();
+  if (!isa_and_nonnull<NonSendableAttr>(nonSendable))
+    return false;
+
+  // GetImplicitSendableRequest::evaluate() creates its extension with the
+  // attribute's AtLoc, so this is a good way to quickly check if the extension
+  // was synthesized for an '@_nonSendable' attribute.
+  return ED->getLocFromSource() == nonSendable->AtLoc;
```
Comment (DougGregor) [inline lib/AST/ASTPrinter.cpp:1787]: This feels a little... brittle... but okay.

File: lib/AST/Attr.cpp [hunk 2]
```diff
@@ -131,6 +131,8 @@ DeclAttrKind DeclAttribute::getAttrKindFromString(StringRef Str) {
 
 /// Returns true if this attribute can appear on the specified decl.
 bool DeclAttribute::canAttributeAppearOnDecl(DeclAttrKind DK, const Decl *D) {
+  if ((getOptions(DK) & OnAnyClangDecl) && D->hasClangNode())
```
Comment (DougGregor) [inline lib/AST/Attr.cpp:134]: This is a clever way to model things!

File: lib/ClangImporter/ClangImporter.cpp [hunk 3]
```diff
@@ -610,6 +618,12 @@ importer::getNormalInvocationArguments(
       "-DSWIFT_CLASS_EXTRA=",
     });
 
+    // Indicate that using '__attribute__((swift_attr))' with '@Sendable' and
+    // '@_nonSendable' on Clang declarations is fully supported, including the
+    // 'attribute push' pragma.
+    if (clangSupportsPragmaAttributeWithSwiftAttr())
+      invocationArgStrs.push_back( "-D__SWIFT_ATTR_SUPPORTS_SENDABLE_DECLS=1");
```
Comment (DougGregor) [inline lib/ClangImporter/ClangImporter.cpp:625]: Does Clang not have a corresponding `__has_feature` we can check in the headers? That's the more traditional way to check for Clang capabilities.

File: lib/ClangImporter/ImporterImpl.h [hunk 4]
```diff
@@ -791,6 +804,31 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
...
+  /// Emit a diagnostic at a clang source location, falling back to a Swift
+  /// location if the clang one is invalid.
+  ///
+  /// The diagnostic will appear in the header file rather than in a generated
+  /// interface. Use this to diagnose issues with declarations that are not
+  /// imported or that are not reflected in a generated interface.
+  template<typename ...Args>
+  void diagnose(HeaderLoc loc, Args &&...args) {
+    // If we're in the middle of pretty-printing, suppress diagnostics.
+    if (SwiftContext.Diags.isPrettyPrintingDecl()) {
```
Comment (DougGregor) [inline lib/ClangImporter/ImporterImpl.h:816]: THANK YOU! I've been tripped up by this a couple of times recently.

**Review by DougGregor — APPROVED** [review]
Some minor comments, but this looks great!

File: lib/ClangImporter/ClangImporter.cpp (hunk 3 above)
Comment (beccadax) [reply lib/ClangImporter/ClangImporter.cpp:625]: Not for saying that a specific attribute is supported by the `attribute push` pragma, no.

File: lib/AST/ASTPrinter.cpp (hunk 1 above)
Comment (beccadax) [reply lib/AST/ASTPrinter.cpp:1787]: Yeah. :/ I couldn’t think of anything better that would be quick, but maybe I can revisit this; arguably we should just print any unavailable Sendable conformance even if it doesn’t come from `@_nonSendable`.

**Pull request merged**